File: swea/D2/1954.달팽이 숫자/1954.달팽이 숫자.cpp
#include <iostream>
#include <vector>

using namespace std;

int T;
int dir[4][2] = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

int main(){
  cin >> T;

  for(int t = 1; t <= T; t++){
    int n;
    cin >> n;

    vector<vector<int>> arr(n, vector<int>(n, 0));
    int x = 0, y = 0, d = 0, cnt = 1;

    for(int i = 0; i < n * n; i++){
      arr[x][y] = cnt;
      
      int nx = x + dir[d][0];
      int ny = y + dir[d][1];

      if((nx >= 0 && nx < n) && (ny >= 0 && ny < n) && arr[nx][ny] == 0){
        x = nx;
        y = ny;
      }
      
      else{
        d = (d + 1) % 4;
        x = x + dir[d][0];
        y = y + dir[d][1];
      }

      cnt ++;
    }

    cout << '#' << t << endl;
    for(int x = 0; x < n; x++){
      for(int y = 0; y < n; y++){
        cout << arr[x][y] << ' ';
      }

      cout << '\n';
    }
  }
}

File: swea/D2/1954.달팽이 숫자/1954.달팽이 숫자.py
import sys
input = sys.stdin.readline

T = int(input())

#하, 우, 상, 좌 순으로 델타 정렬
dx = [0, 1, 0, -1] 
dy = [1, 0, -1, 0]

for a in range(1,T+1):
    n = int(input())
    arr = [[0] * n for _ in range(n)]

    x = 0
    y = 0
    d = 0
    cnt = 1

    #배열 만들었으면 안에서 작업해야지
    for _ in range(n * n):
        
        #탐색한 칸을 칠해
        arr[x][y] = cnt

        #다음 좌표
        nx = x + dx[d]
        ny = y + dy[d]

        #다음 좌표가 불가능한가? 왔던 길 또 가는 건가?
        if 0 <= nx < n and 0 <= ny < n and arr[nx][ny] == 0:
            x = nx
            y = ny
        
        #그렇다면 d값 변환시켜서 다시 도전
        else:
            d = (d + 1) % 4
            x = x + dx[d]
            y = y + dy[d]

        #이동하면 카운트 늘려
        cnt += 1
    
    print(f'#{a}')
    for i in arr:
        print(*i)